### PR TITLE
Add PhoneNumberCast

### PR DESCRIPTION
**Pull request by juliomotol**
This PR aims to add a custom cast (introduced in Laravel 7.x) for `PhoneNumber`.

Given the ff schema:
| users |
| --- |
| id |
| name |
| email |
| contact_number |
| contact_number_country |

You can now do this:

```php
use Propaganistas\LaravelPhone\Models\HasPhoneNumber;
use Propaganistas\LaravelPhone\Models\PhoneNumberCast;

class User extends Model implements HasPhoneNumber
{
    protected $casts = [
        'contact_number' => PhoneNumberCast::Class
    ];

    public function getPhoneNumberCountryColumn($field)
    {
        return 'contact_number_country';
    }
}
```

This also supports multiple phone numbers with different country locale.

| users |
| --- |
| id |
| name |
| email |
| contact_number |
| contact_number_country |
| emergency_number |
| emergency_number_country |

```php
use Propaganistas\LaravelPhone\Models\HasPhoneNumber;
use Propaganistas\LaravelPhone\Models\PhoneNumberCast;

class User extends Model implements HasPhoneNumber
{
    protected $casts = [
        'contact_number' => PhoneNumberCast::Class
        'emergency_number' => PhoneNumberCast::Class
    ];

    public function getPhoneNumberCountryColumn($field)
    {
        switch ($field) {
            case 'contact_number':
                return 'contact_number_country';
            case 'emergency_number'
                return 'emergency_number_country';
        }
    }
}
```

### CODE REVIEW

**Comment by Propaganistas** [comment]
Thank you, this is a good start.

Can you please rework the PR to incorporate the following suggestions?
* I would prefer to use cast parameters instead of an interface and method to retrieve the corresponding country:
```php
protected $casts = [
   'contact_number' => PhoneNumberCast::class.':contact_number_country',
];
```
* The cast parameter should be able to accept one or multiple values, being either model attributes or hardcoded country codes:
```php
protected $casts = [
   'contact_number' => PhoneNumberCast::class.':US,BE,contact_number_country,another_field',
];
```
* I'm not convinced that a country attribute will always be tied only to a phone attribute. Imagine for example some address fields, and the country field is simply reused for phone formatting. I think writing to the country attribute within the Phone Cast might lead to unexpected behavior.
* I think it would be elegant to implement the `Castable` contract in the `PhoneNumber` class.
* Can you add tests for this?

**Comment by juliomotol** [comment]
> * I think writing to the country attribute within the Phone Cast might lead to unexpected behavior.

Yeah, it actually did. During writing the tests, I tried doing:
```
Inquiry::create([
    'mobile_number' => '09123456789',
    'mobile_number_country' => 'PH',
]);
```
 And it was unable to save the `mobile_number_country` within the cast since it hasn't been filled yet due to it being a later index in the array. Your suggestion worked better than what I had initially.

> * I think it would be elegant to implement the `Castable` contract in the `PhoneNumber` class.

Gotcha, I'll have that in as well.

> * Can you add tests for this?

Sure! I'd be happy to.

**Comment by juliomotol** [comment]
One problem with `Castable`, the `castUsing`'s method signature has changed between Laravel [7.x](https://github.com/laravel/framework/blob/7.x/src/Illuminate/Contracts/Database/Eloquent/Castable.php#L12) and [8.x](https://github.com/laravel/framework/blob/8.x/src/Illuminate/Contracts/Database/Eloquent/Castable.php#L14)

File: src/Models/PhoneNumberCast.php [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    protected function resolvePhoneCountry($attributes)
+    {
+        $parsedCountries = $this->parseCountries($this->countries);
+
+        $countryColumns = array_diff($this->countries, $parsedCountries);
+
+        foreach ($countryColumns as $countryColumn) {
+            if ($country = $attributes[$countryColumn]) {
```
Comment (Propaganistas) [inline src/Models/PhoneNumberCast.php:65]: This should be fail-safe, e.g. by using `Arr::get()`

File: tests/Models/User.php [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    protected $fillable = [
+        'contact_number',
+        'contact_number_country',
+        'emergency_number',
+        'emergency_number_country',
+    ];
+
+    protected $casts = [
+        'contact_number' => PhoneNumberCast::class . ':contact_number_country',
+        'emergency_number' => PhoneNumberCast::class . ':emergency_number_country',
```
Comment (Propaganistas) [inline tests/Models/User.php:19]: Can you add a test in which the cast parameters include both a country field and several hardcoded countries?

Comment (juliomotol) [reply tests/Models/User.php:19]: Yup, will be righting more test over the weekend :smile:

File: src/Models/PhoneNumberCast.php (hunk 1 above)
Comment (juliomotol) [reply src/Models/PhoneNumberCast.php:65]: Which part exactly? From what I know, doing `array_diff([1, 2, 3], [1, 2, 3])` will return an `[]`.Giving an empty array in a `foreach`  will do nothing.

Comment (juliomotol) [reply src/Models/PhoneNumberCast.php:65]: I feel like an idiot, did you mean L65? `$attributes[$countryColumn] ?? null`?

Comment (Propaganistas) [reply src/Models/PhoneNumberCast.php:65]: Yes, something like that. It's certainly possible that the country field is not present in `$attributes`.

**Comment by juliomotol** [comment]
Done with the requests and writing test, PR is now open.

**Comment by Propaganistas** [comment]
Thanks! One last suggestion: I don't think the tests need a database setup. A plain model class will suffice:

```php
$user = new User(['phone' => '012 34 56 78']);

// Returns the value of the Cast's `get()`, in this case a PhoneNumber instance
$user->phone;

// Returns the value of the Cast's `set()`, in this case '012 34 56 78'
$user->getAttributes()['phone'];
```

**Comment by juliomotol** [comment]
Ohhh, I see. Well that's handy. I'll apply it later today, thank for the tip!

**Comment by Propaganistas** [comment]
Because of the changing contract from L7 to L8 I'm going to merge this into another branch for now. Will be available in master soon, but this requires some additional thinking first.

**Comment by clarkewing** [comment]
Hi there!

Don't mean to be digging this out prematurely but I'm wondering if you've got any update on when you expect to merge into master? Could I do anything to help?

As a note, in the use cases for my projects, I'll usually store the number in E.164 format in a single column. So far I'd been using accessors, mutators, and overriding the `attributesToArray()` method to serialize properly, and would use `'AUTO'` for the country. Would you be open to defaulting to the `'AUTO'` country if we don't pass any parameters? Like so:
```PHP
protected $casts = [
   'phone' => PhoneNumber::class, // This then implicitly calls PhoneNumber::make($value, 'AUTO')
];
```

**Comment by Propaganistas** [comment]
@clarkewing I've given some love to the `custom-cast` branch. Would you mind checking if the proposal in there floats the boat?

README: https://github.com/Propaganistas/Laravel-Phone/tree/custom-cast#attribute-casting
Code: https://github.com/Propaganistas/Laravel-Phone/compare/custom-cast

**Comment by clarkewing** [comment]
@Propaganistas apologies for the late response, and thanks for looking into it so quickly. I imported the branch into an ongoing project and it works great!

The only thing I would add would be that [we may want to implement the `Illuminate\Contracts\Database\Eloquent\SerializesCastableAttributes` interface](https://laravel.com/docs/8.x/eloquent-mutators#array-json-serialization) so that we can set the serialization behavior.

I'd suggest the following for the `E164PhoneNumberCast`:
```PHP
/**
 * Get the serialized representation of the value.
 *
 * @param  \Illuminate\Database\Eloquent\Model  $model
 * @param  string  $key
 * @param  mixed  $value
 * @param  array  $attributes
 * @return mixed
 */
public function serialize($model, string $key, $value, array $attributes)
{
    return optional($value)->serialize();
}
```

As for the `RawPhoneNumberCast`, while this is not typically something I would use, I assume the following would fit the bill:
```PHP
/**
 * Get the serialized representation of the value.
 *
 * @param  \Illuminate\Database\Eloquent\Model  $model
 * @param  string  $key
 * @param  mixed  $value
 * @param  array  $attributes
 * @return mixed
 */
public function serialize($model, string $key, $value, array $attributes)
{
    return optional($value)-> getRawNumber();
}
```

**Comment by Propaganistas** [comment]
The custom casts have been released in 4.3.0.
Please note that Laravel 8.9 is minimally required.

**Pull request merged**